clippy_lints/src/small_underscore_scope.rs

```diff
@@ -0,0 +1,71 @@
+use crate::utils::{snippet, span_lint_and_sugg};
+use if_chain::if_chain;
+use rustc::lint::{EarlyContext, EarlyLintPass, LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+use rustc_errors::Applicability;
+use syntax::ast::{PatKind, Stmt, StmtKind};
+
+declare_clippy_lint! {
+    /// **What is does:** Checks for underscore bindings inside a struct or tuple which could encompass the entire binding.
+    ///
+    /// **Why is this bad?** The extra binding information is meaningless and makes the code harder to read.
+    ///
+    /// **Known problems:** Won't catch bindings that look like `let (x, (_, _)) = t;`.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// // Bad
+    /// let t = (1, 2);
+    /// let (_, _) = t;
+    ///
+    /// // Good
+    /// let t = (1, 2);
+    /// let _ = t;
+    /// ```
+    pub SMALL_UNDERSCORE_SCOPE,
+    pedantic,
+    "underscore binding occurs inside a struct, but the underscore could be the entire binding"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(SMALL_UNDERSCORE_SCOPE,)
+    }
+
+    fn name(&self) -> &'static str {
+        "SmallUnderscoreScope"
+    }
+}
+
+impl EarlyLintPass for Pass {
+    fn check_stmt(&mut self, cx: &EarlyContext<'_>, stmt: &Stmt) {
+        if_chain! {
+            if let StmtKind::Local(ref local) = stmt.node;
```

Comment:
I think this should apply to all patterns (e.g. in matches, too). You should be able to implement the `check_pat` method instead of the `check_stmt` method